contracts/contracts/VerificationGateway.sol

```diff
@@ -296,12 +296,12 @@ contract VerificationGateway
             IWallet wallet = getOrCreateWallet(bundle.senderPublicKeys[i]);
 
             // check nonce then perform action
-            if (bundle.operations[i].nonce == wallet.nonce()) {
+            if (bundle.operations[i].nonce == wallet.nonce{gas:20000}()) {
```

Comment:
If someone upgrades to a different wallet implementation, its `nonce` function must be runable with 20k gas.
I would add a comment to `IWallet.nonce` to that effect.

Reply:
@jzaki @voltrevo Upon further reflection, hardcoding gas values is an anti-pattern, and could potentially brick the VG (maybe linked wallets?) in the future.

It is not uncommon for gas costs of different ops in the EVM to change over time. Generally they are lowered as hardware gets more powerful and cheaper, but sometimes they are increased for various reasons. See [EIP-2929](https://eips.ethereum.org/EIPS/eip-2929) for an example.

Our BLS contracts from Hubble have to deal with this as well, and are able to [generate the gas costs for BLS sig ops as part of the initialization process](https://github.com/web3well/bls-wallet/blob/main/contracts/contracts/lib/hubble-contracts/contracts/libs/BNPairingPrecompileCostEstimator.sol#L28-L107) . (There also appear to be [some hardcoded values](https://github.com/web3well/bls-wallet/blob/main/contracts/contracts/lib/hubble-contracts/contracts/libs/BNPairingPrecompileCostEstimator.sol#L47-L51), not sure what impact that will have in the future)

Given that, one solution we could do instead would be to to generate & cache this value on init, and then allow someone (public? connected BLS Wallet?) to call a function to recalculate & cache this value if gas costs every change in a way that impacts the VG. There may be other options I'm unaware of as well.

Reply:
@jacque006 Yikes. Yeah that sounds bad.

How about we just include it in the overall gas limit? Something like this:

```solidity=
IWallet wallet = getOrCreateWallet(bundle.senderPublicKeys[i]);

uint256 opGas = bundle.operations[i].gas;
uint256 gasBeforeNonce = gasleft();
uint256 nonce = wallet.nonce{gas:gasRemaining}();
uint256 nonceGas = gasBeforeNonce - gasleft();

// check nonce then perform action
if (bundle.operations[i].nonce == nonce) {
    // request wallet perform operation
    (
        bool success,
        bytes[] memory resultSet
    ) = wallet.performOperation{gas:opGas-nonceGas}(bundle.operations[i]);
    successes[i] = success;
    results[i] = resultSet;
    emit WalletOperationProcessed(
        address(wallet),
        bundle.operations[i].nonce,
        bundle.operations[i].actions,
        successes[i],
        results[i]
    );
}
```

@jzaki wdyt?

Reply:
Will change it to this :+1: This will cater for wallets that are more intensive with their nonce calculation. 
Related #505